src/qml/Legend.qml

```diff
@@ -77,214 +84,279 @@ ListView {
       id: line
       anchors.verticalCenter: parent.verticalCenter
       leftPadding: itemPadding
-      spacing: 5
+      spacing: 0
 
       // Collapsed state visual feedback
-      Rectangle {
+      Row {
+        anchors.verticalCenter: parent.verticalCenter
+        height: 24
+        visible: HasChildren
+
+        Item {
           height: 24
-          width: 10
-          color: "transparent"
+          width: HasChildren ? 24 : 0
+          clip: true
           anchors.verticalCenter: parent.verticalCenter
+          visible: HasChildren
 
           Image {
-              anchors.fill: parent
-              source: isSelectedLayer
-                  ? Theme.getThemeVectorIcon('ic_legend_collapsed_state_white_14dp')
-                  : Theme.getThemeVectorIcon('ic_legend_collapsed_state_14dp')
-              width: 10
-              height: 10
-              rotation: !IsCollapsed ? 90 : 0
-              sourceSize.width: 12 * screen.devicePixelRatio
-              sourceSize.height: 12 * screen.devicePixelRatio
-              fillMode: Image.PreserveAspectFit
-              visible: HasChildren
+            height: parent.height * .75
+            width: height
+            anchors.centerIn: parent
+            source: isSelectedLayer ? whiteIcon : Theme.darkTheme ? whiteIcon : blackIcon
+            rotation: !IsCollapsed ? 90 : 0
+            fillMode: Image.PreserveAspectFit
+            visible: HasChildren
+
+            property var whiteIcon: Theme.getThemeVectorIcon('ic_legend_collapsed_state_white_24dp')
+            property var blackIcon: Theme.getThemeVectorIcon('ic_legend_collapsed_state_24dp')
+
+            Behavior on rotation {
+              NumberAnimation {
+                duration: 100
+              }
+            }
           }
+
+          MouseArea {
+            anchors.fill: parent
+            cursorShape: Qt.PointingHandCursor
+            onClicked: {
+              if (HasChildren) {
+                IsCollapsed = !IsCollapsed
+                projectInfo.saveLayerTreeState()
+              }
+            }
+          }
+        }
+
+        // spacing item only when collapse button available
+        Item {
+          width: 5
+          height: parent.height
+        }
       }
 
-      // Legend image / layer icon
-      Rectangle {
+      Row {
+        anchors.verticalCenter: parent.verticalCenter
+        spacing: 5
+
+        Item {
+          id: layerVisibility
+          height: 24
+          width: visible ? parent.height : 0
+          anchors.verticalCenter: parent.verticalCenter
+          visible: layerTree.data(legend.model.index(index, 0), FlatLayerTreeModel.HasSpatialExtent) === true
```

Comment:
I'm pretty sure you can just do this:

```suggestion
          visible: HasSpatialExtent
```

The reliance on the data() function was merely a requirement in the layer properties window.

Reply:
nice